Merge db-rewrite into master

colagrosso opened the pull request:
See issue #336 for additional discussion.

acabal commented on lib/EbookTag.php:
```diff
@@ -17,4 +22,61 @@ protected function GetUrl(): string{
+	/**
+	 * @throws Exceptions\ValidationException
+	 */
+	public function Validate(): void{
```
Should validate that name exists and is not null or empty. Name should also be trimmed here.

colagrosso replied:
Done, with the same comment about 

```php
$this->Name = trim($this->Name ?? '');
```

above.

acabal commented on lib/Collection.php:
```diff
@@ -1,18 +1,104 @@
+	 * @throws Exceptions\ValidationException
+	 */
+	public function Validate(): void{
+		$error = new Exceptions\ValidationException();
```
Again, validate name

colagrosso replied:
Done.

acabal commented on lib/Ebook.php:
```diff
@@ -306,272 +801,363 @@ public function __construct(?string $wwwFilesystemPath = null){
+	/**
+	 * @throws Exceptions\ValidationException
+	 */
+	public function Validate(): void{
```
Here we must check that the properties you compare are `isset()` first .If they aren't set, then testing against `''` will be a fatal error. You should also trim everything here as well.

colagrosso replied:
Could you clarify something about these properties? It seems there are some that are truly required and marked `NOT NULL` in the DB schema:

* `Identifier`
* `WwwFilesystemPath`
* `RepoFilesystemPath`
* `Title`
* `Description`
* `LongDescription`
* `WordCount`
* `ReadingEase`
* `EbookCreated`
* `EbookUpdated`
* `TextSinglePageByteCount`
* `IndexableText`

For those properties, I trimmed the string ones and checked the others with `isset()`.

Then there are properties that don't seem strictly required and are allowed to be `NULL` in the DB schema:

* `KindleCoverUrl`
* `EpubUrl`
* `AdvancedEpubUrl`
* `KepubUrl`
* `Azw3Url`
* `DistCoverUrl`
* `FullTitle`
* `AlternateTitle`
* `GitHubUrl`
* `WikipediaUrl`

For those, I trimmed them if they are not `null`, but I didn't coerce them from `null` to `''`. That seems right to me.  Some of those URLs don't seem optional to me, though, but maybe formats change over time.

One place where I changed my mind was the `Language` property. It's currently allowed to be `NULL` in the DB, but that seems wrong. I changed the code to not allow a `null` in `$this->Language`, and if you agree I'll update the DB schema, too.

In my 1000+ ebooks, all of them have `Language` set:

```
> select * from Ebooks where Language IS NULL\G
```

acabal replied:
In general, if it's `NOT NULL` in the DB then the property *must* have a value during validation.

Also in general, all string properties should be trimmed during validation.

For properties that *are* nullable, if it's a string, then an empty string should be changed to `null` before saving. This example covers all cases:

````php
function Validate(){
    $this->NullableStringProperty = trim($this->NullableStringProperty ?? '');
    if($this->NullableStringProperty == ''){
        $this->NullableStringProperty = null;
    }

    if(!isset($this->NotNullableProperty)){
        // Validation error - property that is not nullable has not been set before validation.
    }
}
````

Does that answer your question?

colagrosso replied:
> Does that answer your question?

Yep, thanks a bunch for the example code. I that approach to the `Validate()` methods in `Ebook`, `Collection`, `Contributor`, `EbookSource`, `EbookTag`, `GitCommit`, and `LocSubject`. Regarding the "Possible validation error?" for non-nullable string properties, I treated them as validation errors with the same exception as if they property was `null`. That seemed like what we wanted.

acabal commented on lib/Ebook.php:
```diff
@@ -797,23 +1383,318 @@ private function NullIfEmpty($elements): ?string{
+		foreach($this->GitCommits as $commit){
+			$commit->EbookId = $this->EbookId;
+
+			Db::Query('
```
This should be in `GitCommit::Create()` not in `Ebook`

colagrosso replied:
Done.

acabal commented on lib/Ebook.php:
```diff
@@ -797,23 +1383,318 @@ private function NullIfEmpty($elements): ?string{
+				values (?,
+					?,
+				        ?)
+			', [$source->EbookId, $source->Type->value, $source->Url]);
```
This should also be `EbookSource::Create()` not part of `Ebook`

colagrosso replied:
> Pass the enum here, not `->value`

I see you added support for enums in `Db::Query()` a long time ago. Thanks!

> This should also be `EbookSource::Create()` not part of `Ebook`

Done.

acabal commented on lib/Ebook.php:
```diff
@@ -797,23 +1383,318 @@ private function NullIfEmpty($elements): ?string{
+			$contributor->SortOrder = $sortOrder;
+
+			Db::Query('
+				INSERT into Contributors (EbookId, Name, UrlName, SortName, WikipediaUrl, MarcRole, FullName,
```
Again, should be `Contributor::Create()`.

colagrosso replied:
Done.

acabal reviewed (commented):
Some preliminary comments to address before I do a more detailed review

colagrosso commented on lib/EbookTag.php:
```diff
@@ -1,14 +1,19 @@
-		$this->UrlName = Formatter::MakeUrlSafe($this->Name);
-		$this->_Url = '/subjects/' . $this->UrlName;
+	public function __construct(){
+		$this->Type = 'ebook';
```
Done.

colagrosso commented on lib/LocSubject.php:
```diff
@@ -0,0 +1,52 @@
+	/**
+	 * @throws Exceptions\ValidationException
+	 */
+	public function Validate(): void{
```
After adding `trim($this->Name)`, PHPStan throws an error when calling `isset($this->Name)`:

```
 ------ ----------------------------------------------------------------- 
  Line   lib/LocSubject.php                                               
 ------ ----------------------------------------------------------------- 
  14     Property LocSubject::$Name (string) in isset() is not nullable.  
 ------ ----------------------------------------------------------------- 
```

I believe this is due to a change in PHP 8, which also adds a deprecation warning if calling `trim()` on `null`. It looks like a good way to handle the name and avoid the deprecation warning is:

```php
$this->Name = trim($this->Name ?? '');
if($this->Name == ''){
    [...]
}
```

so I did that.

colagrosso commented on lib/Collection.php:
```diff
@@ -1,18 +1,104 @@
+		}
+
+		if($this->Type !== null && ($this->Type != CollectionType::Series && $this->Type != CollectionType::Set)){
+			$error->Add(new Exceptions\InvalidCollectionTypeException($this->Type->value));
```
Done.

acabal commented on lib/Collection.php:
```diff
@@ -1,18 +1,130 @@
+	public ?CollectionType $Type = null;
+	protected ?string $_Url = null;
+
+	protected function GetUrl(): ?string{
```
Return value should not be nullable

colagrosso replied:
Done.

acabal commented on lib/Constants.php:
```diff
@@ -28,10 +28,16 @@
+const EBOOKS_MAX_STRING_LENGTH = 250;
+const EBOOKS_MAX_LONG_STRING_LENGTH = 500;
+const EBOOK_SINGLE_PAGE_SIZE_WARNING = 3 *1024 * 1024; // 3145728
+const EBOOK_EARLIEST_CREATION_DATE = new DateTimeImmutable('2014-01-01');
```
I don't think we need to have this constant or check against it, seems strange to hard code the earliest ebook date.

colagrosso replied:
Done. I thought it would be useful to check for garbage dates that were way too early, e.g., 1970-01-01, but that hasn't been a problem so far.

acabal commented on lib/Ebook.php:
```diff
@@ -797,23 +1535,286 @@ private function NullIfEmpty($elements): ?string{
+	private function InsertLocSubjects(): void{
+		foreach($this->LocSubjects as $locSubject){
+			Db::Query('
+				INSERT into EbookLocSubjects (EbookId, LocSubjectId)
```
Also will throw a duplicate key exception... check all functions in which we add a relation to an ebook for this

colagrosso replied:
Done throughout, with the same caveat above that it's not possible today to throw a `DuplicateDatabaseKeyException`.

acabal commented on lib/EbookSource.php:
```diff
@@ -1,10 +1,61 @@
-	public function __construct(EbookSourceType $type, string $url){
-		$this->Type = $type;
-		$this->Url = $url;
+	public static function FromTypeAndUrl(EbookSourceType $type, string $url): EbookSource{
```
I think this function is unnecessary - if we're creating a new instance, we can just inline the contents of this function instead of having this wrapper because there is no special logic here.

colagrosso replied:
Done.

acabal commented on lib/GitCommit.php:
```diff
@@ -2,21 +2,90 @@
 	 * @throws Exceptions\InvalidGitCommitException
 	 */
-	public function __construct(string $unixTimestamp, string $hash, string $message){
+	public static function FromLog(string $unixTimestamp, string $hash, string $message): GitCommit{
```
I suggest we refactor this to accept a string, since we're saying "create this object from a log line" and not "create this object from several structured parameters". The latter is more akin to `__construct`. Then this function would parse the log line string and create the object. Or, rename this to `__construct`.

colagrosso replied:
Done.

acabal reviewed (changes requested):
OK, I've had some time to do a more in depth review. Basically it looks very good. I think we're almost ready to go. I made some comments on some minor issues, mostly refactoring some names for clarity. Let me know when it's all wrapped up and then we can release it live to see how things go.

acabal commented:
Also, I don't think the `TocEntries` table needs an ID column. We never select on ID and it's a one to many relationship anyway. We *do* order by TocEntryId, but that's wrong - we should have a `SortOrder` column instead, and sort by that.

Likewise for `CollectionEbookId`, `EbookTagId`, `EbookSourceId`, `EbookLocSubjectId`, etc. Those columns are not necessary and if we sort by any of them, we should have a `SortOrder` column instead.

colagrosso commented:
Thanks for the thorough review!

> Let me know when it's all wrapped up and then we can release it live to see how things go.

I'll ping this PR again after I test it all again to make sure it's all wrapped up.

For now, I'm just responding to your comments above. My responses might require another round of review.

colagrosso commented on lib/Ebook.php:
```diff
@@ -306,272 +801,515 @@ public function __construct(?string $wwwFilesystemPath = null){
+		}
+
+		if(isset($this->KindleCoverUrl)){
+			if(!preg_match('|/*_EBOK_portrait.jpg|ius', $this->KindleCoverUrl)){
```
Good catch, done.

colagrosso commented on lib/Ebook.php:
```diff
@@ -306,272 +801,515 @@ public function __construct(?string $wwwFilesystemPath = null){
-		// First the Wikipedia URLs.
-		$this->WikipediaUrl = $this->NullIfEmpty($xml->xpath('/package/metadata/meta[@property="se:url.encyclopedia.wikipedia"][not(@refines)]'));
+		if(isset($this->EpubUrl)){
+			if(!preg_match('|/*.epub|ius', $this->EpubUrl)){
```
Done.

colagrosso commented on lib/Ebook.php:
```diff
@@ -306,272 +801,515 @@ public function __construct(?string $wwwFilesystemPath = null){
+		}
+
+		if(isset($this->AdvancedEpubUrl)){
+			if(!preg_match('|/*_advanced.epub|ius', $this->AdvancedEpubUrl)){
```
Done for all uses of `preg_match`.

colagrosso commented on lib/Ebook.php:
```diff
@@ -306,272 +801,515 @@ public function __construct(?string $wwwFilesystemPath = null){
 
-		$searchString .= ' ' . $this->AlternateTitle;
+		if(isset($this->GitHubUrl)){
+			if(!preg_match('|https://github.com/standardebooks/\w+|ius', $this->GitHubUrl)){
```
Done.

colagrosso commented on lib/Ebook.php:
```diff
@@ -306,272 +801,515 @@ public function __construct(?string $wwwFilesystemPath = null){
+	/**
+	 * @throws Exceptions\ValidationException
+	 */
+	private function InsertTagStrings(): void{
```
Yep, I renamed the functions to follow the new `Ebook::CreateTags()` and `Ebook::AddTags()` pattern.

colagrosso commented on lib/Ebook.php:
```diff
@@ -797,23 +1535,286 @@ private function NullIfEmpty($elements): ?string{
+	private function InsertTags(): void{
+		foreach($this->Tags as $tag){
+			Db::Query('
+				INSERT into EbookTags (EbookId, TagId)
```
I added the check, but I wanted to point out that it's not possible to get a  `DuplicateDatabaseKeyException` from the code as it works today. `Ebook::Save()` calls:

```php  
$this->RemoveTags();  
$this->AddTags();  
```

so any previous entries in `EbookTags` are gone when `AddTags()` runs. I can elaborate on why I like this approach if it would help. Same approach for `LocSubjects`, `CollectionMemberships`, `GitCommits`, `Sources`, `Contributors`, and `TocEntries`.

colagrosso commented on lib/CollectionMembership.php:
```diff
@@ -0,0 +1,15 @@
+<?
+use function Safe\preg_replace;
```
Done.

colagrosso commented on lib/Ebook.php:
```diff
@@ -48,71 +65,574 @@ class Ebook{
+							SELECT *
+							from CollectionEbooks
+							where EbookId = ?
+							order by CollectionEbookId
```
As you pointed out below, I removed `CollectionEbookId` and added a `SortOrder` to this table because that's what we want to sort on, not `SequenceNumber`. The `SortOrder` presents the collection memberships in the same order the producer added them to `content.opf`, e.g. for *To the Lighthouse*:

№ 15 in the Modern Library’s 100 Best Novels set.  
Part of the Encyclopædia Britannica’s Great Books of the Western World set.  
№ 2 in the BBC’s 100 Greatest British Novels (2015) set.

(The sequence numbers there are 15, `NULL`, 2, so they're not what we want to sort on here. We do sort by `SequenceNumber` in `collections/get.php` to present the ebooks in the right order in a collection.)

colagrosso commented on lib/Ebook.php:
```diff
@@ -48,71 +65,574 @@ class Ebook{
+
+			if($index == 0){
+				// No decimal point for smaller than a KB.
+				$this->_TextSinglePageSizeFormatted = sprintf("%d %s", $bytes, $sizes[$index]);
```
I removed the `Safe\sprintf`line above. In PHP8,  `sprintf()` can throw `ValueError` and `ArgumentCountError`:  
   
https://www.php.net/manual/en/function.sprintf.php\#refsect1-function.sprintf-errors

but it's not possible for this code to throw those two exceptions. I didn't add `catch` blocks here, but I will if needed.

colagrosso commented on scripts/update-ebook-database:
```diff
@@ -0,0 +1,55 @@
+try{
+	$ebook->CreateOrUpdate();
+}
+catch(Exceptions\ValidationException $validationException){
```
Running `update-ebook-database --ebookWwwFilesystemPath /does/not/exist` causes an exception in this line above:

```php  
$ebook = Ebook::FromFilesystem($ebookWwwFilesystemPath);    
```

which used to print a fatal error and exit:

```
PHP Fatal error:  Uncaught Exceptions\\EbookNotFoundException: Invalid www filesystem path: /does/not/exist in /standardebooks.org/web/lib/Ebook.php:623  
Stack trace:  
\#0 /standardebooks.org/web/scripts/update-ebook-database(26): Ebook::FromFilesystem()  
\#1 {main}  
  thrown in /standardebooks.org/web/lib/Ebook.php on line 623  
```

but now I added a `try/catch` around that line to print a custom error message:

```
An AppException occurred when updating the ebook database:  
        Exceptions\\EbookNotFoundException: Invalid www filesystem path: /does/not/exist  
```

I think that's what you're asking for. Let me know if I should add the stack trace or any other info to the error message.

The line you commented on catches only `ValidationException`. I'm not expecting any other types of exceptions from `CreateOrUpdate()`, but I could add another `catch` for `AppException` if you think it's a good idea.

colagrosso commented on scripts/update-ebook-database:
```diff
@@ -0,0 +1,55 @@
+}
+
+if($ebookWwwFilesystemPath === null){
+	print("Expected usage: update-ebook-database [-v,--verbose] --ebookWwwFilesystemPath <path>\n");
```
Done.

colagrosso commented:
> Also, I don't think the `TocEntries` table needs an ID column. We never select on ID and it's a one to many relationship anyway. We _do_ order by TocEntryId, but that's wrong - we should have a `SortOrder` column instead, and sort by that.
> 
> Likewise for `CollectionEbookId`, `EbookTagId`, `EbookSourceId`, `EbookLocSubjectId`, etc. Those columns are not necessary and if we sort by any of them, we should have a `SortOrder` column instead.

I removed these 7 primary keys:

* `CollectionEbookId`  
* `ContributorId`  
* `EbookLocSubjectId`  
* `EbookSourceId`  
* `EbookTagId`  
* `GitCommitId`  
* `TocEntryId`

In almost all cases, I added a `SortOrder` column. (`Contributors` already had one. `GitCommits` already sorts on `Created`.)

I believe it's important to sort the collection memberships, tags, sources, etc. by the order the producer listed them in `content.opf`. This matches the site's current behavior.

acabal commented:
OK great! Let me know when you've tested again and you think this is ready to go. I'll start testing then.

colagrosso commented:
Ok, all set! I tested it again on a new droplet, and it worked well. Anecdotally, performance seemed similar to the production site.

There's always room for improvement with my code, but I'm happy stamping this as an initial DB milestone.

acabal commented:
OK, this branch is now live on the site! The site is currently serving ebooks from the DB.

I'll let it shake out for a few days. If everything looks good we can merge this in and I'll make an official announcement!

robinwhittleton commented:
Can confirm that a search for Poe is looking loads better. Thanks for all the hard work!

acabal commented:
I'm coming across an error involving artworks:

If an approved artwork is assigned an SE ebook identifier, but the ebook is not released yet, getting the artwork crashes because the ebook is not in the database.

For example `https://standardebooks.org/artworks/anonymous/down-where-the-glades-begin`.

Here's a stack trace for that URL:

````
PHP Fatal error:  Uncaught Exceptions\EbookNotFoundException: Invalid identifier: url:https://standardebooks.org/ebooks/dashiell-hammett/the-dain-curse in /standardebooks.org/web/lib/Ebook.php:1554
Stack trace:
#0 /standardebooks.org/web/lib/Artwork.php(271): Ebook::GetByIdentifier()
#1 /standardebooks.org/web/lib/Traits/Accessor.php(41): Artwork->GetEbook()
#2 /standardebooks.org/web/templates/ArtworkStatus.php(8): Artwork->__get()
#3 /standardebooks.org/web/lib/Template.php(17): include('...')
#4 /standardebooks.org/web/lib/Template.php(29): Template::Get()
#5 /standardebooks.org/web/www/artworks/get.php(97): Template::__callStatic()
#6 {main}
  thrown in /standardebooks.org/web/lib/Ebook.php on line 1554

colagrosso commented:
Sorry about that. I'll get a fix and test it in a minute.

colagrosso commented:
Ok, sorry about that. I reproduced the issue and confirmed that db3d3b7 fixes it. I pushed that commit to this branch.

acabal commented:
Great, looks like it's fixed. Thanks! 

Looks like there are some merge conflicts, can you take a look when you have a minute?

colagrosso commented:
> Looks like there are some merge conflicts, can you take a look when you have a minute?

Yep, all done.

acabal commented:
OK, it's been a few days and everything looks good. I've merged it in, and everything is now live! Thanks for all your hard work!
